### PR TITLE
Implement calc group score for duplicate to assign replicas evenly

### LINKED ISSUE #4805: Scheduler: support the ability to automatically assign replicas evenly
**What would you like to be added**:
### Background
We want to introduce a new replica assignment strategy in the scheduler, which supports an even assignment of the target replicas across the currently selected clusters.

### Explanation
After going through the filtering, prioritization, and selection phases, three clusters(`member1`, `member2`, `member3`) were selected. We will automatically assign 9 replicas equally among these three clusters, the result we expect is `[{member1: 3}, {member2: 3}, {member3: 3}]`.

**Why is this needed**:
### User Story
As a developer, we have a deployment with 2 replicas that needs to be deployed with high availability across AZs. We hope Karmada can schedule it to two AZs and ensure that there is a replica on each AZ.
2AZ

Our PropagationPolicy might look like this:
```yaml
apiVersion: policy.karmada.io/v1alpha1
kind: PropagationPolicy
metadata:
  name: foo
  namespace: bar
spec:
  resourceSelectors:
    - apiVersion: apps/v1
      kind: Deployment
      name: foo
  placement:
    replicaScheduling:
      replicaSchedulingType: Divided
      replicaDivisionPreference: Weighted
      weightPreference:
        dynamicWeight: AvailableReplicas
    spreadConstraints:
    - spreadByField: zone
      maxGroups: 2
      minGroups: 2
    - spreadByField: cluster
      maxGroups: 2
      minGroups: 2
```
But unfortunately, the strategy `AvailableReplicas` does not guarantee that our replicas are evenly assigned.

### Any ideas?
We can introduce a new replica assignment strategy like `AvailableReplicas`, maybe we can name it `AverageReplicas`. It is essentially different from static weight assignment, because it does not support spread constraints and is mandatory. When assigning replicas, it does not consider whether the cluster can place so many replicas.

### PR DESCRIPTION
**Pull request by ipsum-0320**
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:
Optimize the score calculation method to better measure the pros and cons of the group. This PR implements calc group score for duplicate to assign replicas evenly.

**Which issue(s) this PR fixes**:
Part of #5159

**Special notes for your reviewer**:
@whitewindmills 

**Does this PR introduce a user-facing change?**:
```release-note
`karmada-scheduler`: implement group score calculation instead of take the highest score of clusters
```

### CODE REVIEW

**Comment by whitewindmills** [comment]
wait for https://github.com/karmada-io/karmada/pull/5621
/hold

**Comment by RainbowMango** [comment]
> wait for #5621 /hold

Let's get back on this as #5621 has been merged.

**Comment by ipsum-0320** [comment]
i will resolve the conflicts.

**Comment by ipsum-0320** [comment]
now, i have do something for this PR.  i need you guys on resolving the merge conflict. @whitewindmills @RainbowMango

**Comment by RainbowMango** [comment]
> now, i have do something for this PR. i need you guys on resolving the merge conflict. @whitewindmills @RainbowMango

I can't figure out the intention of this PR. 
And, do you mean you don't know how to resolve the conflicts?
Basically you can do that as follows:
```
1. checkout to your feature branch by `git checkout Implement-group-score-calc-bonus`
2. reset the latest commit on you feature branch by `git reset HEAD~1`
3. stash your commit to cache: `git stash`
4. fetch the upstream and rebase
  * `git fetch upstream`
  * `git rebase upstream/master`
5. restore your commit: `git stash pop` 
  * Note: you can skip this step too if you prefer to write the code from the start. But if you do, you will see the conflicts by checking the status `git status`
 ...
```

**Comment by ipsum-0320** [comment]
> > now, i have do something for this PR. i need you guys on resolving the merge conflict. @whitewindmills @RainbowMango
> 
> I can't figure out the intention of this PR. And, do you mean you don't know how to resolve the conflicts? Basically you can do that as follows:
> 
> ```
> 1. checkout to your feature branch by `git checkout Implement-group-score-calc-bonus`
> 2. reset the latest commit on you feature branch by `git reset HEAD~1`
> 3. stash your commit to cache: `git stash`
> 4. fetch the upstream and rebase
>   * `git fetch upstream`
>   * `git rebase upstream/master`
> 5. restore your commit: `git stash pop` 
>   * Note: you can skip this step too if you prefer to write the code from the start. But if you do, you will see the conflicts by checking the status `git status`
>  ...
> ```

I've resolved the conflicts, so we can consider merging now. Additionally, I'll clarify the purpose of this PR:

As shown in the code, PR #5621 actually represents the GroupScore algorithm when the replica allocation strategy is set to Divided. In contrast, this PR (#5622) provides the GroupScore algorithm for the Duplicate replica allocation strategy. It stands to reason that with Duplicate, the GroupScore algorithm should differ due to the distinct behavior of replica allocation. The GroupScore algorithm I designed works as follows: in a Group, the more clusters that meet the Replica requirement for available copies, the higher the Group Score. If the number of clusters meeting the Replica requirement is the same, then the Group with a higher average cluster score will have a higher Group Score. @whitewindmills @RainbowMango

File: pkg/scheduler/core/spreadconstraint/group_clusters.go [hunk 1]
```diff
@@ -135,10 +135,100 @@ func groupClustersIgnoringTopology(
...
+	// the priority of validClusters is higher than sumValidScore.
+	weightedValidClusters := validClusters * weightUnit
+	return weightedValidClusters + sumValidScore/validClusters
+}
+
 func (info *GroupClustersInfo) calcGroupScore(
 	clusters []ClusterDetailInfo,
 	rbSpec *workv1alpha2.ResourceBindingSpec,
 	minGroups int) int64 {
+	if checkIfDuplicate(rbSpec) {
```
Comment (whitewindmills) [inline pkg/scheduler/core/spreadconstraint/group_clusters.go:227]: we don't need such a redundant function `checkIfDuplicate`, use `rbSpec.Placement.ReplicaSchedulingType() == policyv1alpha1.ReplicaSchedulingTypeDuplicated` instead.

File: pkg/scheduler/core/spreadconstraint/group_clusters.go [hunk 2]
```diff
@@ -135,10 +135,100 @@ func groupClustersIgnoringTopology(
...
+	for _, cluster := range clusters {
+		if cluster.AvailableReplicas >= targetReplica {
+			validClusters++
+			sumValidScore += cluster.Score
+		}
+	}
+
+	// Here is an example, the rbSpec.Replicas == 50.
+
+	// There is the Group 1, it has three clusters as follows.
```
Comment (whitewindmills) [inline pkg/scheduler/core/spreadconstraint/group_clusters.go:160]: five clusters ...

File: pkg/scheduler/core/spreadconstraint/group_clusters.go [hunk 3]
```diff
@@ -135,10 +135,100 @@ func groupClustersIgnoringTopology(
...
+	// we prioritize whether the number of available replicas in each Cluster
+	// meets the Replica requirements. Based on our algorithm, the score for Group 2
+	// is also higher than that of Group 1.
+
+	// Group1's Score = 2 * 1000 + 100 = 2100
+	// Group2's Score = 4 * 1000 + 0 = 4000
+
+	// There is another example, the rbSpec.Replicas == 50.
+
+	// There is the Group 1, it has three clusters as follows.
```
Comment (whitewindmills) [inline pkg/scheduler/core/spreadconstraint/group_clusters.go:191]: ditto

**Review by whitewindmills — COMMENTED** [review]
some nits, LGTM otherwise.

**Comment by ipsum-0320** [comment]
I've solved these problems. @whitewindmills

**Review by RainbowMango — APPROVED** [review]
/approve

**Review by RainbowMango — COMMENTED** [review]
/lgtm

**Pull request merged**